src/backend/gpopt/translate/CQueryMutators.cpp

```diff
@@ -35,6 +35,518 @@ extern "C" {
+		SContextFixGroupDependentTargets ctx_fix_dependent_targets(query);
+		GetVarsWithoutTleWalker((Node *) query->targetList,
+								&ctx_fix_dependent_targets);
+		query->targetList = gpdb::ListConcat(
+			query->targetList, ctx_fix_dependent_targets.m_tlist_addition);
+
+		// Step 2.
+		// Store current unique TargetEntry references in groupClause.
+		// After modifying groupClause, they will be required to update
+		// arguments inside grouping functions.
+		SContexGetGroupUniqueTleWalker ctx_old_grouping_tle_refs;
+		GetGroupUniqueTleReferencesWalker((Node *) query->groupClause,
+										  &ctx_old_grouping_tle_refs);
+
+		// Step 3.
+		// Update groupClause - add all functionally dependent entries explicitly.
+		ListCell *lc_constraint = NULL;
+		ForEach(lc_constraint, query->constraintDeps)
+		{
+			Oid constraint_oid = lfirst_oid(lc_constraint);
+			Oid conrelid = 0;
+			Oid confrelid =
+				0;	// not used, but function get_constraint_relation_oids requires it..
+			gpdb::GetConstraintRelationOids(constraint_oid, &conrelid,
+											&confrelid);
+
+			ctx_fix_dependent_targets.m_conrelid = conrelid;
+			ctx_fix_dependent_targets.m_conkeys =
+				gpdb::GetConstraintRelationColumns(constraint_oid);
+
+			ListCell *lc_tle = NULL;
+			ForEach(lc_tle, query->targetList)
+			{
+				TargetEntry *target_entry = (TargetEntry *) lfirst(lc_tle);
+				Node *expr = (Node *) target_entry->expr;
+
+				if (IsA(expr, Var))
+				{
+					Var *var = (Var *) expr;
+					RangeTblEntry *rte = (RangeTblEntry *) gpdb::ListNth(
+						query->rtable, var->varno - 1);
+					GPOS_ASSERT(NULL != rte);
+
+					// skip relations that are not constrained by primary key, that is
+					// currently being checked
+					if (conrelid != rte->relid)
+					{
+						continue;
+					}
+
+					if (!CTranslatorUtils::IsGroupingColumn(
+							expr, query->groupClause, query->targetList))
+					{
+						// add new SortGroupClause to groupClause
+						SortGroupClause *gc;
+						Oid sortop;
+						Oid eqop;
+						bool hashable;
+
+						gpdb::GetSortGroupOperators(exprType(expr), true, true,
+													false, &sortop, &eqop, NULL,
+													&hashable);
+
+						gc = makeNode(SortGroupClause);
+						gc->tleSortGroupRef = gpdb::AssignSortGroupRef(
+							target_entry, query->targetList);
+						gc->eqop = eqop;
+						gc->sortop = sortop;
+						gc->nulls_first = false;
+						gc->hashable = hashable;
+
+						ctx_fix_dependent_targets.m_gc = gc;
+						ctx_fix_dependent_targets.m_parent_is_grouping_clause =
+							false;
+
+						List *old_group_clause = query->groupClause;
+
+						query->groupClause =
+							(List *) AddMissingGroupClauseMutator(
+								(Node *) query->groupClause,
+								&ctx_fix_dependent_targets);
+
+						gpdb::ListFreeDeep(old_group_clause);
+						gpdb::GPDBFree(gc);
+					}
+				}
+			}
+			gpdb::ListFree(ctx_fix_dependent_targets.m_conkeys);
+		}
+		// remove constraintDeps
+		gpdb::ListFree(query->constraintDeps);
```

Comment:
why do we remove the list? In order to avoid the repeating of the algorithm during dxl translation?

Reply:
The constraintDeps list is filled in check_functional_grouping in case it detects that there is a TargetEntry that is not in groupClause, but the groupClause has primary key (thus the TargetEntry is functionally dependent on the grouping columns). 
So, ```SELECT a, b FROM test_table1 GROUP BY a;``` will have not empty constraintDeps list, while ```SELECT a, b FROM test_table1 GROUP BY a, b;``` will have empty constraintDeps  list.
As in FixGroupDependentTargets we eliminate this situation by  adding columns to groupClause (we can say that ```SELECT a, b FROM test_table1 GROUP BY a;``` is transformed to ```SELECT a, b FROM test_table1 GROUP BY a, b;```), I think it will be correct to cleanup the list.

Reply:
What if ORCA fails after this line? For example, a query like this:
```sql
create table t1(a int primary key, b varchar(255), c int);
create table t2(a int primary key, b varchar(25), c int);
explain (costs off)
select * from t1 join t2 using (b) group by t1.a, t2.a
```
In this case, nothing bad will happen because the postgres planner doesn't call the find_expr_references_walker function, but what happens if it does?